contrib/filter-extensions/README.md

```diff
@@ -0,0 +1,141 @@
+provides a simple example, and `filter_sparse()` is more complex.
+
+The API differences between the built-in filters and the filter extensions:
+
+1. Filter extensions don't handle `omitset`s directly, instead setting `omit`.
+2. Filter extensions receive a void pointer they can use for context.
+
+## Building
+
+There is some plumbing in the Git Makefile to help with this via
+`FILTER_EXTENSIONS`, setting it to space-separated paths of the filter extension
+static libraries indicates that these filters should be compiled into git.
+For example:
+
+```console
+FILTER_EXTENSIONS="contrib/filter-extensions/rand/rand.a" make
+```
+
+Filter extensions don't need to be within the Git source tree. A filter extension
+static library should either exist at the given path - ie, `rand.a` should exist -
+or there should be a Makefile in that directory which will create it when
+`make rand.a` is run. (Such a Makefile should also have a `clean` target which
+deletes all object files and brings the directory back to its initial state).
+
+The static library should define a struct of type `filter_extension` called
+`filter_extension_NAME` where `NAME` is the name of your extension (ie `rand`
+for `rand.a`). See
+[list-objects-filter-extensions.h](../../list-objects-filter-extensions.h)
+
+This definition should follow the following pattern:
+```C
+#include "list-objects-filter-extensions.h"
+
+/* Definitions of rand_init, rand_filter_object, rand_free ... */
+
+const struct filter_extension filter_extension_rand = {
+    "rand",
+    &rand_init,
+    &rand_filter_object,
+    &rand_free,
+};
+```
+
+(The names of your `init_fn`, `filter_object_fn` and `free_fn` are not important,
+but the string literal should again be the the name of your extension - `"rand"`
+for the filter extension in `rand.a`.)
+
+You may use library functions from Git if you include the relevant Git headers,
+since the filter extensions and Git itself will be linked together into a single
+binary.
+
+You may depend on other libraries if you indicate that they are to be linked
+into the Git binary using `FILTER_EXTENSION_LIBS`. See the C++ example below.
+
+## Developing in C++ (and other languages)
+
+You can develop filter extensions with C++, but many Git header files are not
+compatible with modern C++, so you won't be able to directly use Git library
+functions. However, you can use them if you create wrapper functions in C that
+delegates to the Git library functions you need, but which are also C++
+compatible. See [`rand_cpp`](./rand_cpp/) for a simple example. A similar
+solution would be to implement the extension itself in C, and have the extension
+do any operations that require Git library functions, but have it delegate to a
+C wrapper API that you add to a C++ library that already contains the
+domain-specific operations that you need. In either case, remember to wrap any
+functions that must be C-compatible with `extern C` when declaring or
+defining them from within C++.
```

Comment:
I added a crude runtime initialisation test to `rand.cpp` for testing:

```patch
diff --git a/contrib/filter-extensions/rand_cpp/rand.cpp b/contrib/filter-extensions/rand_cpp/rand.cpp
index a907e999cb..4d27bd85d5 100644
--- a/contrib/filter-extensions/rand_cpp/rand.cpp
+++ b/contrib/filter-extensions/rand_cpp/rand.cpp
@@ -9,6 +9,13 @@ extern "C" {
        #include "adapter_functions.h"
 }

+// run-time dynamic initialisation test
+static int hello() {
+       std::cerr << "filter-rand-cpp: good morning\n";
+       return 123;
+}
+static int A_NUMBER = hello();
+
 struct rand_context {
        int percentageMatch = 0;
        int matchCount = 0;
```

This works (`gcc` == `clang` on macOS by default)
```console
$ make CC=clang LDFLAGS=-lc++ FILTER_EXTENSIONS=contrib/filter-extensions/rand_cpp/rand_cpp.a
$ ./git --version
filter-rand-cpp: good morning
git version 2.32.0.rc1.753.g680de0773f.dirty
$ ./git rev-list refs/heads/master --objects --max-count 1 --filter=extension:rand_cpp=3 --filter-print-omitted | grep -c '^~'
filter-rand-cpp: good morning
filter-rand-cpp: matching 3%
filter-rand-cpp: done: count=4057 (blob=3856 tree=201) matched=117 elapsed=0.008535s rate=475337/s average=2.10377us
3739
```
This doesn't link properly:
```console
$ make CC=gcc-11 LDFLAGS=-lstdc++ FILTER_EXTENSIONS=contrib/filter-extensions/rand_cpp/rand_cpp.a
```

Reply:
Here's a basic patch to enable using an alternate compiler linker driver for linking (eg: `CC_LD=g++` / `CC_LD=clang++` / `CC_LD=g++-11`). 

```patch
diff --git a/Makefile b/Makefile
index 5fbf2de14e..db26a0444d 100644
--- a/Makefile
+++ b/Makefile
@@ -552,6 +552,7 @@ export prefix bindir sharedir sysconfdir gitwebdir perllibdir localedir

 # Set our default programs
 CC = cc
+CC_LD = $(CC)
 AR = ar
 RM = rm -f
 DIFF = diff
@@ -2240,7 +2241,7 @@ git.sp git.s git.o: EXTRA_CPPFLAGS = \
        '-DGIT_INFO_PATH="$(infodir_relative_SQ)"'

 git$X: git.o GIT-LDFLAGS $(BUILTIN_OBJS) $(GITLIBS) $(EXTENSION_LIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) \
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) \
                $(filter %.o,$^) $(LIBS)

 help.sp help.s help.o: command-list.h
@@ -2604,17 +2605,17 @@ compat/nedmalloc/nedmalloc.sp: SP_EXTRA_FLAGS += -Wno-non-pointer-null
 endif

 git-%$X: %.o GIT-LDFLAGS $(GITLIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) $(LIBS)
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) $(LIBS)

 git-imap-send$X: imap-send.o $(IMAP_SEND_BUILDDEPS) GIT-LDFLAGS $(GITLIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
                $(IMAP_SEND_LDFLAGS) $(LIBS)

 git-http-fetch$X: http.o http-walker.o http-fetch.o GIT-LDFLAGS $(GITLIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
                $(CURL_LIBCURL) $(LIBS)
 git-http-push$X: http.o http-push.o GIT-LDFLAGS $(GITLIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
                $(CURL_LIBCURL) $(EXPAT_LIBEXPAT) $(LIBS)

 $(REMOTE_CURL_ALIASES): $(REMOTE_CURL_PRIMARY)
@@ -2624,7 +2625,7 @@ $(REMOTE_CURL_ALIASES): $(REMOTE_CURL_PRIMARY)
        cp $< $@

 $(REMOTE_CURL_PRIMARY): remote-curl.o http.o http-walker.o GIT-LDFLAGS $(GITLIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
                $(CURL_LIBCURL) $(EXPAT_LIBEXPAT) $(LIBS)

 $(LIB_FILE): $(LIB_OBJS)
@@ -2930,7 +2931,7 @@ perf: all
 t/helper/test-tool$X: $(patsubst %,t/helper/%,$(TEST_BUILTINS_OBJS))

 t/helper/test-%$X: t/helper/test-%.o GIT-LDFLAGS $(GITLIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) $(filter %.a,$^) $(LIBS)
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) $(filter %.a,$^) $(LIBS)

 check-sha1:: t/helper/test-tool$X
        t/helper/test-sha1.sh
```

These both work:
```console
$ make CC=clang CC_LD=clang++ FILTER_EXTENSIONS=contrib/filter-extensions/rand_cpp/rand_cpp.a
$ ./git --version
filter-rand-cpp: good morning
git version 2.32.0.rc1.753.g680de0773f.dirty
$ ./git rev-list refs/heads/master --objects --max-count 1 --filter=extension:rand_cpp=3 --filter-print-omitted | grep -c '^~'
filter-rand-cpp: good morning
filter-rand-cpp: matching 3%
filter-rand-cpp: done: count=4057 (blob=3856 tree=201) matched=117 elapsed=0.010891s rate=372509/s average=2.6845us
3739

$ make CC=gcc-11 CC_LD=g++-11 FILTER_EXTENSIONS=contrib/filter-extensions/rand_cpp/rand_cpp.a
$ ./git --version
filter-rand-cpp: good morning
git version 2.32.0.rc1.753.g680de0773f.dirty
$ ./git rev-list refs/heads/master --objects --max-count 1 --filter=extension:rand_cpp=3 --filter-print-omitted | grep -c '^~'
filter-rand-cpp: good morning
filter-rand-cpp: matching 3%
filter-rand-cpp: done: count=4057 (blob=3856 tree=201) matched=117 elapsed=0.007292s rate=556363/s average=1.79739us
3739
```

Reply:
So I guess need to check it on Linux and under MSVC?

Reply:
... I wonder how much effort we should go to, to have a C++ filter extension in contrib/ and in the documentation? Mostly, we want to people to be able to a) contribute filters written in C in a straightforward manner or b) use a language of their choice, and use their specific knowledge of that language's C bindings to debug any linking issues. Having a C++ example I suppose is a good way to help people think through the general problem of b), and it's mostly working, so I guess we'll fix it up so it 100% works?

Not yet done, I'll apply your patch in a bit

Reply:
I guess we submit it and see what comments are like, it can always live as an example outside Git's repo if need be.  Feels like if people want to use filters to do Clever Things™ then that could quite often involve some external libraries. If there's a straightforward path to tread that'll certainly help them.